src/dev/lupio/lupio_pic.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_pic.hh"
+
+#include "cpu/base.hh"
+#include "debug/LupioPIC.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioPIC.hh"
+#include "sim/system.hh"
+
+#define LUPIO_PIC_NSRC      32
+
+namespace gem5
+{
+
+using namespace RiscvISA;
+
+LupioPIC::LupioPIC(const Params &params) :
+    BasicPioDevice(params, params.pio_size),
+    system(params.system),
+    nSrc(params.n_src),
+    nContext(params.n_contexts)
+{
+    DPRINTF(LupioPIC, "LupioPIC initalized\n");
+}
+
+void
+LupioPIC::post(int src_id)
+{
+    gem5_assert(src_id < nSrc && src_id >= 0);
+
+    uint32_t irq_mask = 1UL << src_id;
+    pending |= irq_mask;
+    lupioPicUpdateIRQ();
+}
+
+void
+LupioPIC::clear(int src_id)
+{
+    gem5_assert(src_id < nSrc && src_id >= 0);
+
+    uint32_t irq_mask = 1UL << src_id;
+    pending &= ~irq_mask;
+    lupioPicUpdateIRQ();
+}
+
+void
+LupioPIC::lupioPicUpdateIRQ()
+{
+    for (int i = 0; i < nContext; i++) {
+        int thread_id = i >> 1;
+        int int_id = (i & 1) ?
+            ExceptionCode::INT_EXT_SUPER : ExceptionCode::INT_EXT_MACHINE;
```

Comment:
Don't copy from the PLIC, their approach is different and very RISC-V oriented (ie this notion of contexts).

For now, I would strongly suggest making a mono-cpu design, that is N sources (devices connected to PIC) and only 1 output (the CPU).

The type of interrupt (eg int_ext_super) should be a parameter to the device.